src/NLog/Layouts/Typed/Layoutable.cs

```diff
@@ -0,0 +1,288 @@
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.ComponentModel;
+using System.Globalization;
+using System.Reflection;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout with a simple value (e.g. int) or a layout which results in a simple value (e.g. ${counter})
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class Layoutable<T> : TypedLayout<T>
+    {
+        /// <inheritdoc />
+        public Layoutable(T value) : base(value)
+        {
+        }
+
+        /// <inheritdoc />
+        public Layoutable(Layout layout) : base(layout)
+        {
+        }
+
+        #region Overrides of TypedLayout<T>
+
+        /// <inheritdoc />
+        protected override string TypedName => typeof(T).Name; //todo cache?
+
+        /// <inheritdoc />
+        protected override string ValueToString(T value, CultureInfo cultureInfo)
+        {
+            if (value is IConvertible convertible)
+            {
+                return convertible.ToString(cultureInfo);
+            }
+
+            return value?.ToString();
+        }
+
+        /// <inheritdoc />
+        protected override bool TryParse(string text, out T value)
+        {
+            return TryConvertTo(text, out value);
+        }
+
+        /// <inheritdoc />
+        protected override bool TryConvertTo(object raw, out T value)
+        {
+            value = default; // todo check
+            if (raw == null)
+            {
+                return true;
+            }
+
+            try
+            {
+                // todo could be better?
+                var objTypeCode = (default(T) as IConvertible)?.GetTypeCode();
+                switch (objTypeCode)
+                {
+                    //todo ugly :P
+                    case TypeCode.Boolean:
+                        {
+                            if (bool.TryParse(raw.ToString(), out var value1))
+                            {
+                                value = (T)(object)value1;
```

Comment:
Is this a box/unbox operation?

Reply:
Related https://stackoverflow.com/questions/1189144/c-sharp-non-boxing-conversion-of-generic-enum-to-int#23391746